internal/keppel/security_scan_policy.go

```diff
@@ -0,0 +1,187 @@
+/******************************************************************************
+*
+*  Copyright 2023 SAP SE
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+******************************************************************************/
+
+package keppel
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/sapcc/go-bits/errext"
+	"github.com/sapcc/go-bits/regexpext"
+
+	"github.com/sapcc/keppel/internal/clair"
+	"github.com/sapcc/keppel/internal/trivy"
+)
+
+// SecurityScanPolicy is a policy enabling user-defined adjustments to
+// vulnerability reports generated by Trivy.
+type SecurityScanPolicy struct {
+	//NOTE: We have code that uses slices.Contains() to locate policies. Be careful
+	//when adding fields that cannot be meaningfully compared with the == operator.
+	ManagingUserName          string                   `json:"managed_by_user,omitempty"`
+	RepositoryRx              regexpext.BoundedRegexp  `json:"match_repository"`
+	NegativeRepositoryRx      regexpext.BoundedRegexp  `json:"except_repository,omitempty"`
+	VulnerabilityIDRx         regexpext.BoundedRegexp  `json:"match_vulnerability_id"`
+	NegativeVulnerabilityIDRx regexpext.BoundedRegexp  `json:"except_vulnerability_id,omitempty"`
+	ExceptFixReleased         bool                     `json:"except_fix_released,omitempty"`
+	Action                    SecurityScanPolicyAction `json:"action"`
+}
+
+// SecurityScanPolicyAction appears in type SecurityScanPolicy.
+type SecurityScanPolicyAction struct {
+	Assessment string                    `json:"assessment"`
+	Ignore     bool                      `json:"ignore,omitempty"`
+	Severity   clair.VulnerabilityStatus `json:"severity,omitempty"`
+}
+
+// String returns the JSON representation of this policy (for use in log and
+// error messages).
+func (p SecurityScanPolicy) String() string {
+	//we only obtain SecurityScanPolicy instances through unmarshaling, so it is
+	//safe to assume that they will marshal without error
+	buf, err := json.Marshal(p)
+	if err != nil {
+		panic(err.Error())
```

Comment:
Shouldn't we return some contex here? Otherwise the error message would be hard to track down.

Reply:
Probably something about the securityScanPolicy.

If the database entry is somehow bit rotten, we could get into an error here.

Reply:
That would fail on **unmarshal**, not on marshal.